hedera-node/src/main/java/com/hedera/services/tokens/HederaTokenStore.java

```diff
@@ -143,12 +157,63 @@ public boolean isCreationPending() {
 
 	@Override
 	public void setHederaLedger(HederaLedger hederaLedger) {
+		hederaLedger.setTokenRelsLedger(tokenRelsLedger);
 		this.hederaLedger = hederaLedger;
 	}
 
 	@Override
-	public void setLedger(TransactionalLedger<AccountID, AccountProperty, MerkleAccount> ledger) {
-		this.ledger = ledger;
+	public void setAccountsLedger(TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accountsLedger) {
+		this.accountsLedger = accountsLedger;
+	}
+
+	@Override
+	public ResponseCodeEnum associate(AccountID aId, List<TokenRef> tokens) {
+		return fullySanityChecked(aId, tokens, (account, tokenIds) -> {
+			var accountTokens = hederaLedger.getAssociatedTokens(aId);
+			for (TokenID id : tokenIds) {
+				if (accountTokens.includes(id)) {
+					return TOKEN_ALREADY_ASSOCIATED_TO_ACCOUNT;
+				}
+			}
+			int effectiveRelationships = accountTokens.purge(id -> !exists(id), id -> get(id).isDeleted());
+			var validity = OK;
+			if ((effectiveRelationships + tokenIds.size()) > properties.maxTokensPerAccount()) {
+				validity = TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED;
+			} else {
+				accountTokens.associateAll(new HashSet<>(tokenIds));
+				for (TokenID id : tokenIds) {
+					var relationship = asTokenRel(aId, id);
+					tokenRelsLedger.create(relationship);
+					var token = get(id);
+					tokenRelsLedger.set(
+							relationship,
+							TokenRelProperty.IS_FROZEN,
+							token.hasFreezeKey() && token.accountsAreFrozenByDefault());
+					tokenRelsLedger.set(
+							relationship,
+							TokenRelProperty.IS_KYC_GRANTED,
+							!token.hasKycKey());
+				}
+			}
+			hederaLedger.setAssociatedTokens(aId, accountTokens);
+			return validity;
+		});
+	}
+
+	@Override
+	public ResponseCodeEnum dissociate(AccountID aId, List<TokenRef> tokens) {
```

Comment:
I think that we should update the Total Supply of the tokens as-well when performing disassociation.